File: credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // ---------- 1. Get input ----------
    long long number = get_long("Number: ");

    // ---------- 2. Luhn checksum ----------
    int sum = 0;
    int position = 0;          // 0 = last digit, 1 = second-to-last, …
    long long n = number;      // copy so we can destroy it

    while (n > 0)
    {
        int digit = n % 10;    // current right-most digit

        if (position % 2 == 1) // every other digit, starting with second-to-last
        {
            int doubled = digit * 2;
            sum += doubled / 10 + doubled % 10; // add the two digits of the product
        }
        else
        {
            sum += digit;
        }

        n /= 10;
        position++;
    }

    bool valid = (sum % 10 == 0);

    // ---------- 3. Determine length and leading digits ----------
    int length = position;

    // first two digits
    long long leading = number;
    while (leading >= 100)            // reduce until only two digits remain
    {
        leading /= 10;
    }
    int first_two = (int) leading;
    int first_one = first_two / 10;

    // ---------- 4. Identify issuer ----------
    if (valid)
    {
        if (length == 15 && (first_two == 34 || first_two == 37))
        {
            printf("AMEX\n");
            return 0;
        }
        if (length == 16 && (first_two >= 51 && first_two <= 55))
        {
            printf("MASTERCARD\n");
            return 0;
        }
        if ((length == 13 || length == 16) && first_one == 4)
        {
            printf("VISA\n");
            return 0;
        }
    }

    // Anything that reaches here is invalid
    printf("INVALID\n");
}

File: python/credit.py
def get_card_number() -> int:
    """Prompt until the user types a positive integer."""
    while True:
        num_str = input("Number: ").strip()
        if num_str.isdigit():          # all characters 0-9
            return int(num_str)

def luhn_valid(number: int) -> bool:
    """Return True if `number` passes the Luhn checksum."""
    total = 0
    for pos, digit_char in enumerate(reversed(str(number))):
        digit = int(digit_char)

        if pos % 2 == 1:               # every second digit from the right
            doubled = digit * 2
            total += doubled // 10 + doubled % 10   # add the digits of the product
        else:
            total += digit

    return total % 10 == 0

def issuer(number: int) -> str:
    """Identify the card issuer or return 'INVALID'."""
    num_str   = str(number)
    length    = len(num_str)
    first_one = int(num_str[0])
    first_two = int(num_str[:2])

    if luhn_valid(number):
        if length == 15 and first_two in (34, 37):
            return "AMEX"
        if length == 16 and 51 <= first_two <= 55:
            return "MASTERCARD"
        if length in (13, 16) and first_one == 4:
            return "VISA"

    return "INVALID"

def main() -> None:
    number = get_card_number()
    print(issuer(number))

if __name__ == "__main__":
    main()
